src/platform/packages/shared/kbn-zod-helpers/src/deep_strict.ts

```diff
@@ -7,321 +7,54 @@
- * composed entirely of parsable schemas)
- */
-export function assertAllParsableSchemas(
-  outerSchema: z.Schema
-): Set<{ key: string; schema: DangerousParsableSchema }> {
-  // track processed schemas, to prevent self-referencing schemas
-  // instantiated with z.lazy() to create recursion errors
-  const processed: Set<Schema> = new Set();
-  // track dangerous schemas and their path
-  const dangerous: Set<{ key: string; schema: DangerousParsableSchema }> = new Set();
-
-  innerAssertAllParsableSchemas(outerSchema, '');
-
-  return dangerous;
-
-  function innerAssertAllParsableSchemas(schema: z.Schema, key: string): void {
-    assertIsParsableSchema(schema, key);
-
-    if (processed.has(schema)) {
-      return;
-    }
-
-    processed.add(schema);
-
-    if (isPrimitiveParsableSchema(schema)) {
-      return;
-    }
-
-    if (isDangerousParsableSchema(schema)) {
-      dangerous.add({ key, schema });
-      return;
-    }
-
-    const def = schema._def;
-
-    switch (def.typeName) {
-      case ZodFirstPartyTypeKind.ZodObject:
-        return Object.entries(def.shape()).forEach(([prop, innerSchema]) =>
-          innerAssertAllParsableSchemas(innerSchema as z.Schema, key ? `${key}.${prop}` : prop)
-        );
-      case ZodFirstPartyTypeKind.ZodOptional:
-        return innerAssertAllParsableSchemas(def.innerType, key);
-      case ZodFirstPartyTypeKind.ZodArray:
-        return innerAssertAllParsableSchemas(def.type, key);
-
-      case ZodFirstPartyTypeKind.ZodUnion:
-      case ZodFirstPartyTypeKind.ZodDiscriminatedUnion:
-        const schemas: z.Schema[] = def.options;
-        return schemas.forEach((innerSchema) => innerAssertAllParsableSchemas(innerSchema, key));
-
-      case z.ZodFirstPartyTypeKind.ZodIntersection:
-        innerAssertAllParsableSchemas(def.left, key);
-        innerAssertAllParsableSchemas(def.right, key);
-        return;
-
-      case z.ZodFirstPartyTypeKind.ZodEffects:
-        if (def.effect.type === 'transform') {
-          dangerous.add({ key, schema: schema as ZodEffects<ZodAny> });
-        }
-        return innerAssertAllParsableSchemas(def.schema, key);
-
-      case z.ZodFirstPartyTypeKind.ZodRecord:
-        return innerAssertAllParsableSchemas(def.valueType, key);
-
-      case ZodFirstPartyTypeKind.ZodDefault:
-        return innerAssertAllParsableSchemas(def.innerType, key);
-
-      case ZodFirstPartyTypeKind.ZodLazy:
-        return innerAssertAllParsableSchemas(def.getter(), key);
-
-      case ZodFirstPartyTypeKind.ZodNullable:
-        return innerAssertAllParsableSchemas(def.innerType, key);
-    }
+import { z, ZodIssueCode } from '@kbn/zod';
+import { difference, isArray, isPlainObject } from 'lodash';
+
+function getFlattenedKeys(obj: unknown, parentKey = '', keys: Set<string> = new Set()) {
+  if (isPlainObject(obj)) {
+    Object.entries(obj as Record<string, any>).forEach(([key, value]) => {
+      getFlattenedKeys(value, parentKey ? `${parentKey}.${key}` : key, keys);
```

Comment:
Considering potential optimizations it's possible to apply backtracking approach for the `parentKey`. Since there is a single thread and we have to walk through the all candidates `parentKey` could be an array

```ts
function getFlattenedKeys(obj: unknown, parentKeys: string[] = [], keys: Set<string> = new Set()) {
  if (isPlainObject(obj)) {
    Object.entries(obj as Record<string, any>).forEach(([key, value]) => {
      parentKeys.push(key);
      getFlattenedKeys(value, parentKeys, keys);
      parentKeys.pop();
    });
  } else if (isArray(obj)) {
    obj.forEach((value) => {
      getFlattenedKeys(value, parentKeys, keys);
    });
  } else {
    keys.add(parentKeys.join('.'));
  }
  return keys;
}
```